semantic-core/src/Data/Core/Pretty.hs

```diff
@@ -56,57 +55,58 @@ inParens amount go = do
-  If (Const con) (Const tru) (Const fal) -> do
-    con' <- "if"   `appending` con
-    tru' <- "then" `appending` tru
-    fal' <- "else" `appending` fal
-    pure $ Pretty.sep [con', tru', fal']
-
-  Load (Const p)   -> "load" `appending` p
-  Edge Lexical (Const n) -> "lexical" `appending` n
-  Edge Import (Const n) -> "import" `appending` n
-  Const item :. Const body   -> inParens 4 $ do
-    f <- item
-    g <- body
-    pure (f <> symbol "." <> g)
-
-  Const lhs := Const rhs -> inParens 3 $ do
-    f <- lhs
-    g <- rhs
-    pure (f <+> symbol "=" <+> g)
-
-  -- Annotations are not pretty-printed, as it lowers the signal/noise ratio too profoundly.
-  Ann _ (Const c) -> c
-  where bind (Ignored x) f = let x' = name x in (,) x' <$> local (x':) (getConst (unScope f))
+prettyCore :: Style -> Term Core User -> AnsiDoc
+prettyCore style = run . runReader @Prec 0 . go (pure . name)
+  where go :: (Member (Reader Prec) sig, Carrier sig m) => (a -> m AnsiDoc) -> Term Core a -> m AnsiDoc
+        go var = \case
+          Var v -> var v
+          Term t -> case t of
+            Let a -> pure $ keyword "let" <+> name a
+            a :>> b -> do
+              prec <- ask @Prec
+              fore <- with 12 (go var a)
+              aft  <- with 12 (go var b)
+
+              let open  = symbol ("{" <> softline)
+                  close = symbol (softline <> "}")
+                  separator = ";" <> Pretty.line
+                  body = fore <> separator <> aft
+
+              pure . Pretty.align $ encloseIf (12 > prec) open close (Pretty.align body)
+
+            Lam n f -> inParens 11 $ do
+              (x, body) <- bind n f
+              pure (lambda <> x <+> arrow <+> body)
+
+            Frame    -> pure $ primitive "frame"
+            Unit     -> pure $ primitive "unit"
+            Bool b   -> pure $ primitive (if b then "true" else "false")
+            String s -> pure . strlit $ Pretty.viaShow s
+
+            f :$ x -> inParens 11 $ (<+>) <$> go var f <*> go var x
+
+            If con tru fal -> do
+              con' <- "if"   `appending` go var con
+              tru' <- "then" `appending` go var tru
+              fal' <- "else" `appending` go var fal
+              pure $ Pretty.sep [con', tru', fal']
+
+            Load p   -> "load" `appending` go var p
+            Edge Lexical n -> "lexical" `appending` go var n
+            Edge Import n -> "import" `appending` go var n
+            item :. body   -> inParens 4 $ do
+              f <- go var item
+              g <- go var body
+              pure (f <> symbol "." <> g)
+
+            lhs := rhs -> inParens 3 $ do
+              f <- go var lhs
+              g <- go var rhs
+              pure (f <+> symbol "=" <+> g)
+
+            -- Annotations are not pretty-printed, as it lowers the signal/noise ratio too profoundly.
+            Ann _ c -> go var c
+          where bind (Ignored x) f = let x' = name x in (,) x' <$> go (incr (const (pure x')) var) (fromScope f)
```

Comment:
This is the real meat of the change: we use `fromScope` to transform the `Scope () f a` (which is isomorphic to `f (Incr () (f a))`) to `f (Incr () a)`, and then call `go` to recur over it with an extended continuation. I.e. if `var :: a -> m AnsiDoc`, then `incr (const (pure x')) var :: Incr () a -> m AnsiDoc`. Using this continuation enables us to avoid performing _n²_ `fmap`s as we recur through the structure.